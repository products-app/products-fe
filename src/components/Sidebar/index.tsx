import { X as CloseIcon } from 'phosphor-react'
import { useState } from 'react'
import classNames from 'classnames'

const styles = {
  sidebar:
    'fixed top-0 right-0 h-screen w-96 overflow-y-auto z-50 p-8 pt-12 bg-gray900 block transition ease-in-out delay-350 duration-[1000ms]',
  sidebarOpen: 'translate-x-0',
  sidebarClose: 'translate-x-full',
  btnClose: 'absolute top-0 right-0 pt-10 pr-10',
}

type SidebarProps = {
  isOpen?: boolean
  children: React.ReactNode
}

const Sidebar = ({ isOpen, children }: SidebarProps) => {
  const [open, setOpen] = useState(isOpen)

  const handleClose = () => {
    setOpen(!open)
  }

  return (
    <div
      className={classNames(styles.sidebar, {
        [styles.sidebarClose]: !open,
        [styles.sidebarOpen]: open,
      })}
    >
      <button onClick={handleClose} className={styles.btnClose}>
        <CloseIcon className="text-white text-lg" />
      </button>
      {children}
    </div>
  )
}

export default Sidebar
