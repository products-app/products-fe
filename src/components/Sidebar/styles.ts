import { styled } from '@lebernardo/react'

export default {
  sidebar:
    'fixed top-0 right-0 h-screen w-96 overflow-y-auto z-[9000] px-8 pt-12 bg-gray900 block transition ease-in-out duration-[1000ms] h-full max-h-full overflow-hidden box-border',
  sidebarOpen: 'translate-x-0',
  sidebarClose: 'translate-x-full',
}

export const SidebarContainer = styled('section', {
  background: '$gray900',
  position: 'fixed',
  top: 0,
  right: 0,
  height: '100vh',
  width: '24rem',
  overflowY: 'auto',
  zIndex: 9000,
  paddingLeft: '3rem 2rem 0',
  display: 'block',
  transitionProperty: 'color, background-color, border-color, text-decoration-color, fill, stroke, opacity, box-shadow, transform, filter, backdrop-filter',
  transitionTimingFunction: 'cubic-bezier(0.4, 0, 0.2, 1)',
  transitionDuration: '1000ms',
  maxHeight: '100%',
  overflow: 'hidden',
  boxSizing: 'border-box',
})

export const ButtonClose = styled('div', {
  position: 'absolute',
  top: 0,
  right: 0,
  paddingTop: '2.5rem',
  paddingRight: '2.5rem',
  zIndex: 9999,
  cursor: 'pointer',

  svg: {
    color: '$white',
    fontSize: '$lg',
  }
});