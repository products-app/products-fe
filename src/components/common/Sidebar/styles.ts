import { styled } from '@lebernardo/react'

export const SidebarContainer = styled('section', {
  background: '$gray900',
  position: 'fixed',
  top: 0,
  right: 0,
  height: '100vh',
  width: '24rem',
  overflowY: 'auto',
  zIndex: 9000,
  paddingLeft: '3rem 2rem 0',
  display: 'block',
  transitionProperty:
    'color, background-color, border-color, text-decoration-color, fill, stroke, opacity, box-shadow, transform, filter, backdrop-filter',
  transitionTimingFunction: 'cubic-bezier(0.4, 0, 0.2, 1)',
  transitionDuration: '1000ms',
  maxHeight: '100%',
  overflow: 'hidden',
  boxSizing: 'border-box',
})

export const ButtonClose = styled('button', {
  position: 'absolute',
  top: 0,
  right: 0,
  paddingTop: '2.5rem',
  paddingRight: '2.5rem',
  zIndex: 9999,
  cursor: 'pointer',
  border: 'none',

  svg: {
    color: '$white',
    fontSize: '$lg',
  },
})
