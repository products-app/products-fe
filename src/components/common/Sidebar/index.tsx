import { X as CloseIcon } from 'phosphor-react'
import { SidebarContainer, ButtonClose } from './styles'

type SidebarProps = {
  isOpen?: boolean
  children?: React.ReactNode
  onClose: () => void
}

const Sidebar = ({ isOpen, children, onClose }: SidebarProps) => {
  let style
  if (isOpen) {
    style = { transform: 'translateX(0px)' }
  } else {
    style = { transform: 'translateX(100%)' }
  }

  return (
    <SidebarContainer id="sidebar" style={style}>
      <ButtonClose aria-label="button-close-sidebar" onClick={onClose}>
        <CloseIcon />
      </ButtonClose>
      {children}
    </SidebarContainer>
  )
}

const defaultProps = {
  isOpen: false,
  children: <></>,
}

Sidebar.defaultProps = defaultProps
export default Sidebar
