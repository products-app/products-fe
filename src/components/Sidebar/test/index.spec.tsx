import { render } from '@testing-library/react'
import Sidebar from '../index'

const content = <div>Test</div>
const props = {
  onClose: () => {},
  isOpen: false,
}

describe('Sidebar component', () => {
  it('should render correctly', () => {
    const { container } = render(<Sidebar {...props} />)
    expect(container).toMatchSnapshot()
  })

  it('should contain text in sidebar content', () => {
    const { getByText } = render(
      <Sidebar {...props} isOpen={true}>
        {content}
      </Sidebar>,
    )
    expect(getByText('Test')).toBeDefined()
    expect(() => getByText('no exist')).toThrow()
  })
})
