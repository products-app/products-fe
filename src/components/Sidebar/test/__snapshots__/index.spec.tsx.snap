// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Sidebar component > should render correctly 1`] = `
<div>
  <div
    class="fixed top-0 right-0 h-screen w-96 overflow-y-auto z-[9000] px-8 pt-12 bg-gray900 block transition ease-in-out duration-[1000ms] h-full max-h-full overflow-hidden box-border translate-x-full"
  >
    <button
      class="absolute top-0 right-0 pt-10 pr-10 z-[9999]"
    >
      <svg
        class="text-white text-lg"
        fill="currentColor"
        height="1em"
        viewBox="0 0 256 256"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <rect
          fill="none"
          height="256"
          width="256"
        />
        <line
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="16"
          x1="200"
          x2="56"
          y1="56"
          y2="200"
        />
        <line
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="16"
          x1="200"
          x2="56"
          y1="200"
          y2="56"
        />
      </svg>
    </button>
  </div>
</div>
`;
